File: src/TopiaForge.ModManager.Core/LaunchPlan.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace TopiaForge.ModManager.Core
{
    public sealed class LaunchRequest
    {
        public LaunchRequest(string targetId, string? worldOverride = null, string? transitionOverride = null)
        {
            TargetId = LaunchContractValues.Identifier(targetId, nameof(targetId));
            WorldOverride = worldOverride == null ? null
                : LaunchContractValues.Identifier(worldOverride, nameof(worldOverride));
            TransitionOverride = transitionOverride == null ? null
                : LaunchContractValues.Choice(transitionOverride, nameof(transitionOverride), ModTransitions.ByPrecedence);
        }

        public string TargetId { get; }
        public string? WorldOverride { get; }
        public string? TransitionOverride { get; }
    }

    /// <summary>Immutable, serializable claim about a resolution. Runtime re-resolution establishes authority.</summary>
    public sealed class LaunchPlanDescriptor
    {
        public LaunchPlanDescriptor(string targetId, string gamemodeId, string worldId, string transition,
            LaunchRequest request, IEnumerable<PackageIdentity> packages, string? worldFamilyId = null, string? digest = null)
        {
            TargetId = LaunchContractValues.Identifier(targetId, nameof(targetId));
            GamemodeId = LaunchContractValues.Identifier(gamemodeId, nameof(gamemodeId));
            WorldId = LaunchContractValues.Identifier(worldId, nameof(worldId));
            WorldFamilyId = LaunchContractValues.OptionalIdentifier(worldFamilyId, nameof(worldFamilyId));
            Transition = LaunchContractValues.Choice(transition, nameof(transition), ModTransitions.ByPrecedence);
            Request = new LaunchRequest(request.TargetId, request.WorldOverride, request.TransitionOverride);
            Packages = LaunchContractValues.Packages(packages);
            Digest = PackageSetDigest.Of(Packages);
            if (digest != null && LaunchContractValues.Digest(digest) != Digest)
                throw new ArgumentException("The descriptor digest does not match its package identities.", nameof(digest));
            if (!string.Equals(TargetId, Request.TargetId, StringComparison.OrdinalIgnoreCase))
                throw new ArgumentException("The resolved target must match the original request.", nameof(targetId));
            if (WorldFamilyId != null && (!WorldId.StartsWith(WorldFamilyId + ".", StringComparison.OrdinalIgnoreCase)
                || WorldId.Length <= WorldFamilyId.Length + 1))
                throw new ArgumentException("The resolved world must be a concrete instance of its family.");
        }

        public string TargetId { get; }
        public string GamemodeId { get; }
        public string WorldId { get; }
        public string? WorldFamilyId { get; }
        public string Transition { get; }
        public LaunchRequest Request { get; }
        public IReadOnlyList<PackageIdentity> Packages { get; }
        public string Digest { get; }
    }

    /// <summary>Resolved authority plus private declaration snapshots; getters never expose owned mutable DTOs.</summary>
    public sealed class LaunchPlan
    {
        private readonly ModManifest declarations;

        internal LaunchPlan(LaunchPlanDescriptor descriptor, ModLaunchTargetDeclaration target,
            ModGamemodeDeclaration gamemode, ModWorldDeclaration world)
        {
            Descriptor = descriptor;
            declarations = ModManifestJson.CopyForLaunch(new ModManifest
            {
                Contributions = new ModContributions
                {
                    LaunchTargets = new List<ModLaunchTargetDeclaration> { target },
                    Gamemodes = new List<ModGamemodeDeclaration> { gamemode },
                    Worlds = new List<ModWorldDeclaration> { world }
                }
            });
        }

        public LaunchPlanDescriptor Descriptor { get; }
        public string TargetId => Descriptor.TargetId;
        public string GamemodeId => Descriptor.GamemodeId;
        public string WorldId => Descriptor.WorldId;
        public string? WorldFamilyId => Descriptor.WorldFamilyId;
        public string Transition => Descriptor.Transition;
        public LaunchRequest Request => Descriptor.Request;
        public IReadOnlyList<PackageIdentity> Packages => Descriptor.Packages;
        public string Digest => Descriptor.Digest;
        public ModLaunchTargetDeclaration Target => ModManifestJson.CopyForLaunch(declarations).Contributions!.LaunchTargets[0];
        public ModGamemodeDeclaration Gamemode => ModManifestJson.CopyForLaunch(declarations).Contributions!.Gamemodes[0];
        public ModWorldDeclaration World => ModManifestJson.CopyForLaunch(declarations).Contributions!.Worlds[0];
    }

    /// <summary>The established FNV-1a package-set consistency digest, independent of JSON ordering.</summary>
    public static class PackageSetDigest
    {
        public static string Of(IReadOnlyList<ResolvedPackage> packages) => Of(packages.Select(package => package.Identity));
        public static string Of(IEnumerable<PackageIdentity> packages) => OfCanonical(packages
            .Select(package => package.Id + "@" + package.Version).OrderBy(value => value, StringComparer.Ordinal));

        public static string OfCanonical(IEnumerable<string> canonicalEntries)
        {
            var hash = 14695981039346656037UL;
            var first = true;
            foreach (var entry in canonicalEntries)
            {
                if (!first) hash = Mix(hash, (byte)'\n');
                first = false;
                foreach (var character in entry)
                {
                    hash = Mix(hash, (byte)(character & 0xff));
                    hash = Mix(hash, (byte)((character >> 8) & 0xff));
                }
            }

            return hash.ToString("x16", CultureInfo.InvariantCulture);
        }

        private static ulong Mix(ulong hash, byte value) => unchecked((hash ^ value) * 1099511628211UL);
    }
}

File: packages/launcher_domain/lib/src/launch_resolution/launch_plan.dart
part of '../launch_resolution.dart';

final class LaunchRequest {
  LaunchRequest({
    required String targetId,
    String? worldOverride,
    String? transitionOverride,
  }) : targetId = _declaration(targetId),
       worldOverride = worldOverride == null
           ? null
           : _declaration(worldOverride),
       transitionOverride = transitionOverride == null
           ? null
           : _choice(transitionOverride, ModTransitions.byPrecedence);
  factory LaunchRequest.fromJson(Object? value) {
    final json = _object(
      value,
      {'targetId', 'worldOverride', 'transitionOverride'},
      {'targetId'},
    );
    return LaunchRequest(
      targetId: _declaration(json['targetId']),
      worldOverride: json.containsKey('worldOverride')
          ? _declaration(json['worldOverride'])
          : null,
      transitionOverride: json.containsKey('transitionOverride')
          ? _choice(json['transitionOverride'], ModTransitions.byPrecedence)
          : null,
    );
  }
  final String targetId;
  final String? worldOverride;
  final String? transitionOverride;
  Map<String, Object?> toJson() => {
    'targetId': targetId,
    if (worldOverride != null) 'worldOverride': worldOverride,
    if (transitionOverride != null) 'transitionOverride': transitionOverride,
  };
}

/// Inactive transport descriptor; every launch is re-resolved before execution.
final class LaunchPlanDescriptor {
  LaunchPlanDescriptor({
    required String targetId,
    required String gamemodeId,
    required String worldId,
    String? worldFamilyId,
    required String transition,
    required LaunchRequest request,
    required Iterable<PackageIdentity> packages,
    String? digest,
  }) : targetId = _declaration(targetId),
       gamemodeId = _declaration(gamemodeId),
       worldId = _declaration(worldId),
       worldFamilyId = worldFamilyId == null
           ? null
           : _declaration(worldFamilyId),
       transition = _choice(transition, ModTransitions.byPrecedence),
       request = LaunchRequest.fromJson(request.toJson()),
       packages = _identities(packages) {
    if (!_idEquals(this.targetId, this.request.targetId) ||
        (this.worldFamilyId != null &&
            (!this.worldId.toLowerCase().startsWith(
                  '${this.worldFamilyId!.toLowerCase()}.',
                ) ||
                this.worldId.length <= this.worldFamilyId!.length + 1))) {
      throw const FormatException('Inconsistent launch identities.');
    }
    this.digest = packageSetDigest(this.packages);
    if (digest != null && _digest(digest) != this.digest) {
      throw const FormatException('Package digest does not match identities.');
    }
  }
  factory LaunchPlanDescriptor.fromJson(Object? value) {
    final json = _object(
      value,
      {
        'targetId',
        'gamemodeId',
        'worldId',
        'worldFamilyId',
        'transition',
        'request',
        'packages',
        'digest',
      },
      {
        'targetId',
        'gamemodeId',
        'worldId',
        'transition',
        'request',
        'packages',
        'digest',
      },
    );
    return LaunchPlanDescriptor(
      targetId: _declaration(json['targetId']),
      gamemodeId: _declaration(json['gamemodeId']),
      worldId: _declaration(json['worldId']),
      worldFamilyId: json.containsKey('worldFamilyId')
          ? _declaration(json['worldFamilyId'])
          : null,
      transition: _choice(json['transition'], ModTransitions.byPrecedence),
      request: LaunchRequest.fromJson(json['request']),
      packages: _list(json['packages'], PackageIdentity.fromJson),
      digest: _digest(json['digest']),
    );
  }
  final String targetId;
  final String gamemodeId;
  final String worldId;
  final String? worldFamilyId;
  final String transition;
  final LaunchRequest request;
  final List<PackageIdentity> packages;
  late final String digest;
  String get launchTargetId => targetId;
  List<PackageIdentity> get resolvedPackages => packages;
  Map<String, Object?> toJson() => {
    'targetId': targetId,
    'gamemodeId': gamemodeId,
    'worldId': worldId,
    if (worldFamilyId != null) 'worldFamilyId': worldFamilyId,
    'transition': transition,
    'request': request.toJson(),
    'packages': packages.map((item) => item.toJson()).toList(),
    'digest': digest,
  };
}

/// Only the resolver creates authority, with copied selected declarations.
final class LaunchPlan extends LaunchPlanDescriptor {
  LaunchPlan._({
    required super.targetId,
    required super.gamemodeId,
    required super.worldId,
    super.worldFamilyId,
    required super.transition,
    required super.request,
    required super.packages,
    required ModLaunchTargetDeclaration target,
    required ModGamemodeDeclaration gamemode,
    required ModWorldDeclaration world,
  }) : _targetJson = jsonEncode(target.toJson()),
       _gamemodeJson = jsonEncode(gamemode.toJson()),
       _worldJson = jsonEncode(world.toJson());
  final String _targetJson;
  final String _gamemodeJson;
  final String _worldJson;
  ModLaunchTargetDeclaration get target =>
      ModLaunchTargetDeclaration.fromJson(jsonDecode(_targetJson));
  ModGamemodeDeclaration get gamemode =>
      ModGamemodeDeclaration.fromJson(jsonDecode(_gamemodeJson));
  ModWorldDeclaration get world =>
      ModWorldDeclaration.fromJson(jsonDecode(_worldJson));
  LaunchPlanDescriptor get descriptor => LaunchPlanDescriptor(
    targetId: targetId,
    gamemodeId: gamemodeId,
    worldId: worldId,
    worldFamilyId: worldFamilyId,
    transition: transition,
    request: request,
    packages: packages,
    digest: digest,
  );
}

/// The established FNV-1a digest over sorted id@version UTF-16LE code units.
/// Entries are separated by one LF byte. This is a consistency check only.
String packageSetDigest(Iterable<PackageIdentity> packages) {
  final entries = packages.map((item) => '${item.id}@${item.version}').toList()
    ..sort();
  return packageSetDigestOfCanonical(entries);
}

String packageSetDigestOfCanonical(List<String> entries) {
  var hash = BigInt.parse('14695981039346656037');
  final prime = BigInt.parse('1099511628211');
  final mask = (BigInt.one << 64) - BigInt.one;
  void mix(int byte) {
    hash = ((hash ^ BigInt.from(byte)) * prime) & mask;
  }

  for (var i = 0; i < entries.length; i++) {
    if (i > 0) mix(10);
    for (final unit in entries[i].codeUnits) {
      mix(unit & 255);
      mix((unit >> 8) & 255);
    }
  }
  return hash.toRadixString(16).padLeft(16, '0');
}
